sql/migrate/lex.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package migrate
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"unicode/utf8"
+)
+
+// stmts provides a generic implementation for extracting
+// SQL statements from the given file string.
+func stmts(input string) ([]string, error) {
+	var stmts []string
+	l, err := newLex(input)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		s, err := l.stmt()
+		if err == io.EOF {
+			return stmts, nil
+		}
+		if err != nil {
+			return nil, err
+		}
+		// Trim new lines between statements.
+		stmts = append(stmts, strings.TrimSpace(s))
+	}
+}
+
+type lex struct {
+	input string
+	pos   int    // current position.
+	width int    // size of latest rune.
+	start int    // start position of current statement.
+	depth int    // depth of parentheses.
+	delim string // configured delimiter.
+}
+
+const (
+	eos          = -1
+	delimiter    = ";"
+	delimComment = "-- atlas:delimiter"
+)
+
+func newLex(input string) (*lex, error) {
+	delim := delimiter
+	input = strings.TrimSpace(input)
+	if strings.HasPrefix(input, delimComment) {
+		input = input[len(delimComment):]
+		if !strings.HasPrefix(input, " ") {
+			return nil, fmt.Errorf("expect space after %q, got: %s", delimComment, input[:1])
+		}
+		input = strings.TrimSpace(input)
+		i := strings.Index(input, "\n")
+		if i == -1 {
+			return nil, errors.New("empty delimiter")
+		}
+		delim = strings.NewReplacer(`\n`, "\n", `\r`, "\r", `\t`, "\t").
+			Replace(input[:i])
+		input = strings.TrimSpace(input[i+1:])
```

Comment:
Add some comments here about what is happening please.

Reply:
When scanning a delimiter string like `\n`, we get an escaped one. i.e. 2 bytes, `\\` and `n`.

Reply:
I know. Still not easy to read and this is a good example of where a comment brings great value to a reviewer or when touching the code again.

Reply:
Agree, I already updated the code before commenting here.